website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+
+- **Structured Responses**: Delivers outputs in a predefined format, including:
+
+- A **description** of the query's purpose.
+
+- The corresponding **SQL code** ready for execution.
+
+- **Adaptable and Reliable**: Uses GPT-4 for robust, consistent query generation, minimizing manual effort and errors.
+
+This chain is a critical component in our workflow, enabling seamless integration of SQL query generation with downstream processes, ensuring accuracy, and significantly improving efficiency.
+
+```python
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_openai import ChatOpenAI
+from pydantic import BaseModel, Field
+
+class SQLQuery(BaseModel):
+    """Schema for SQL query solutions to questions."""
+    description: str = Field(description="Description of the SQL query")
+    sql_code: str = Field(description="The SQL code block")
+
+def get_sql_gen_chain():
+    """Set up the SQL generation chain."""
+    sql_gen_prompt = ChatPromptTemplate.from_messages(
+        [
+            (
+                "system",
+                """You are an SQL assistant with expertise in SQL query generation. \n
+Answer the user's question based on the provided documentation snippets and the database schema provided below. Ensure any SQL query you provide is valid and executable. \n
+Structure your answer with a description of the query, followed by the SQL code block. Here are the documentation snippets:\n{retrieved_docs}\n\nDatabase Schema:\n
+- Customers(CustomerID, CustomerName, ContactName, Address, City, PostalCode, Country)
+- Orders(OrderID, CustomerID, OrderDate)
+- OrderDetails(OrderDetailID, OrderID, ProductID, Quantity)
+- Products(ProductID, ProductName, Price)\n
+Here is the user's question:""",
+            ),
+            ("placeholder", "{messages}"),
+        ]
+    )
+
+    # Initialize the OpenAI LLM
+    llm = ChatOpenAI(temperature=0, model="gpt-4")
+
+    # Create the code generation chain
+    sql_gen_chain = sql_gen_prompt | llm.with_structured_output(SQLQuery)
+
+    return sql_gen_chain
+```
+
+### **Node Descriptions**
+#### **1. Translate Input**
+
+The `translate_input` node translates user queries into English to standardize processing, ensuring all inputs are compatible with downstream nodes.
+
+**Purpose:** Normalize user input to English for consistent processing.
+
+**Process:**
+
+- Extracts the latest user input from the state.
+
+- Uses the `ChatOpenAI` model to translate the input or verify it’s already in English.
+
+- Updates the state with the translated input.
+
+**Key Considerations:**
+
+- Maintains original intent and context during translation.
+
+- Provides accurate results even for multilingual queries.
+
+**Examples:**
+
+- Input: _"Quantos pedidos foram realizados em Novembro?"_
+
+- Translated: _"How many orders were made in November?"_
+
+- Input: _"Combien de ventes avons-nous enregistrées en France ?"_
+
+- Translated: _"How many sales did we record in France?"_
```

Comment:
Note early on that multilingual refers to taking natural-language inputs in multiple languages, not e.g. supporting multiple SQL dialects or something like that. I wasn't 100% sure until I got to here!

Reply:
Thanks for pointing this out, it makes a lot of sense! I've clarified early on that the term 'multi-lingual' refers specifically to the system's ability to process natural language inputs in multiple languages, not to supporting multiple SQL dialects. I hope this resolves any potential ambiguity.